### PR TITLE
Bookish books copies

### PR DESCRIPTION
**Pull request by liliannaszabo**

### CODE REVIEW

File: app/controllers/copies_controller.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+class CopiesController < ApplicationController
+  def index
+    @book = Book.find(params[:book_id])
```
Comment (felix-waller-softwire) [inline app/controllers/copies_controller.rb]: What is `@book` used for?

File: app/controllers/copies_controller.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    @copy = @book.copies.create(comment_params)
+    redirect_to book_path(@book)
+  end
+
+
+
+  def comment_params
+    params.require(:copy).permit(:borrower, :due_date)
+  end
```
Comment (felix-waller-softwire) [inline app/controllers/copies_controller.rb:20]: Should this be `copy_params`?

File: app/controllers/copies_controller.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
+class CopiesController < ApplicationController
+  def index
+    @book = Book.find(params[:book_id])
+    @copies = @book.copies
+  end
+  def show
+    @copy = params
```
Comment (felix-waller-softwire) [inline app/controllers/copies_controller.rb]: Could you link to how this works?

File: app/views/books/_form.html.erb [hunk 4]
```diff
@@ -0,0 +1,29 @@
+<%= form_with model: @book do |form| %>
+  <div>
+    <%= form.label :title, "Title" %>
+    <%= form.text_field :title, required: true, placeholder: "Enter book title" %>
```
Comment (felix-waller-softwire) [inline app/views/books/_form.html.erb]: Placeholder text is usually used as an example of what the input should contain, rather than additional or supplemental instructions.

Since some screen readers don't announce placeholders and since in these cases the placeholders replicate the labels, I'd suggest removing them.

Some sources say placeholders are harmful due to their low contrast or ability to make it seem like a field is already filled (especially if you increase the contrast), W3C has some [example research](https://www.w3.org/WAI/GL/low-vision-a11y-tf/wiki/Placeholder_Research)

File: app/views/books/_form.html.erb [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    <% @book.errors.full_messages_for(:title).each do |message| %>
+      <div><%= message %></div>
```
Comment (felix-waller-softwire) [inline app/views/books/_form.html.erb]: You might want to look into [aria-describedby](https://www.w3.org/WAI/WCAG21/Techniques/aria/ARIA21#example-2-identifying-errors-in-data-format) for associating each error with the relevant field

File: app/views/books/index.html.erb [hunk 6]
```diff
@@ -0,0 +1,10 @@
+<h1>My Books</h1>
+<ul>
+  <% @books.sort_by(&:title).each do|book| %>
+    <li>
+      <a href="/books/<%= book.id %>">
+        <%= book.title %>
+      </a>
```
Comment (felix-waller-softwire) [inline app/views/books/index.html.erb]: Rails has a `link_to` component that is more declarative: https://guides.rubyonrails.org/routing.html#generating-paths-and-urls-from-code

**Review by felix-waller-softwire — COMMENTED** [review]
Looks good :)

File: app/views/books/index.html.erb (hunk 6 above)
Comment (liliannaszabo) [reply app/views/books/index.html.erb]: Changed it now 👍

File: app/controllers/copies_controller.rb (hunk 3 above)
Comment (liliannaszabo) [reply app/controllers/copies_controller.rb]: I don't think it does actually. Will change it now

File: app/controllers/copies_controller.rb (hunk 1 above)
Comment (liliannaszabo) [reply app/controllers/copies_controller.rb]: Probably better idea to just chain them

File: app/controllers/copies_controller.rb (hunk 2 above)
Comment (liliannaszabo) [reply app/controllers/copies_controller.rb:20]: Nice catch, renamed now

File: app/models/book.rb [hunk 7]
```diff
@@ -0,0 +1,6 @@
+class Book < ApplicationRecord
+  has_many :copies
+  validates :title, presence: true
+  validates :author, presence: true
+  validates :isbn, presence: true
```
Comment (dbarnettS) [inline app/models/book.rb:5]: Another validation could be to ensure that the isbn is numerical https://guides.rubyonrails.org/active_record_validations.html#numericality

**Review by dbarnettS — APPROVED** [review]
Looks all good!

File: app/controllers/books_controller.rb [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+    if @book.save
+      redirect_to @book
+    else
+      render :new, status: :unprocessable_entity
+    end
+  end
+
+
+  def edit
```
Comment (8kristy) [inline app/controllers/books_controller.rb:24]:
```suggestion
  end

  def edit
```

File: app/controllers/books_controller.rb [hunk 9]
```diff
@@ -0,0 +1,48 @@
+class BooksController < ApplicationController
+  def index
+    @books = Book.all
+  end
+
+  def show
+    @book = Book.find(params[:id])
+  end
+  def new
```
Comment (8kristy) [inline app/controllers/books_controller.rb:9]:
```suggestion
  end
  
  def new
```

File: app/controllers/books_controller.rb [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+  end
+
+  def destroy
+    @book = Book.find(params[:id])
+    @book.destroy
+    redirect_to root_path, status: :see_other
+  end
+
+
+  def book_params
```
Comment (8kristy) [inline app/controllers/books_controller.rb:45]:
```suggestion
  end

  def book_params
```

File: app/controllers/copies_controller.rb [hunk 11]
```diff
@@ -0,0 +1,22 @@
+class CopiesController < ApplicationController
+  def index
+    @copies = Book.find(params[:book_id]).copies
+  end
+  def show
+    @copy = Book.find(params[:book_id]).copies.find(params[:id])
+  end
+
+
+  def create
```
Comment (8kristy) [inline app/controllers/copies_controller.rb:10]:
```suggestion
  end

  def create
```

File: app/controllers/copies_controller.rb [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+    @book = Book.find(params[:book_id])
+    @copy = @book.copies.create(comment_params)
+    redirect_to book_path(@book)
+  end
+
+
+
+  def comment_params
```
Comment (8kristy) [inline app/controllers/copies_controller.rb:18]:
```suggestion
  end

  def comment_params
```

File: app/helpers/books_helper.rb [hunk 13]
```diff
@@ -0,0 +1,2 @@
+module BooksHelper
```
Comment (8kristy) [inline app/helpers/books_helper.rb:1]: What are these helpers for?

File: app/views/books/_form.html.erb [hunk 14]
```diff
@@ -0,0 +1,29 @@
+<%= form_with model: @book do |form| %>
+  <div>
+    <%= form.label :title, "Title" %>
+    <%= form.text_field :title, required: true, aria: { describedby: 'title-error' } %>
+    <% @book.errors.full_messages_for(:title).each do |message| %>
+      <div id="title-error"><%= message %></div>
+    <% end %>
+  </div>
```
Comment (8kristy) [inline app/views/books/_form.html.erb:8]: These fields look very similar, might be worth generalising them into a separate template and import?

File: app/views/books/show.html.erb [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+    Borrower: <%= copy.borrower %>
+  </p>
+  <p>
+    Due Date: <%= copy.due_date %>
+  </p>
+  <p>__</p>
+<% end %>
+
+
+<h2>Add a copy:</h2>
```
Comment (8kristy) [inline app/views/books/show.html.erb:23]:
```suggestion
<% end %>

<h2>Add a copy:</h2>
```

File: app/views/books/show.html.erb [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+<ul>
+  <li><%= link_to "Edit", edit_book_path(@book) %></li>
+  <li><%= link_to "Destroy", book_path(@book), data: {
+    turbo_method: :delete,
+    turbo_confirm: "Are you sure?"
+  } %></li>
+  
+</ul>
+
+Copies
```
Comment (8kristy) [inline app/views/books/show.html.erb:11]: Does this need to be a header?

File: app/views/books/show.html.erb [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+<% @book.copies.each do |copy| %>
```
Comment (8kristy) [inline app/views/books/show.html.erb:12]: Not necessary but if you added more sections like this (especially if they appear on other pages) it would be worth to break them out into a separate file - both for resuing and readability

File: app/views/copies/index.html.erb [hunk 18]
```diff
@@ -0,0 +1,6 @@
+<h1>My Books</h1>
+<ul>
+  <% for copy in @copies %>
+    <li><%= copy.borrower %></li>
```
Comment (8kristy) [inline app/views/copies/index.html.erb:4]: Why are we listing the borrower for "my books"?

File: test/controllers/books_controller_test.rb [hunk 19]
```diff
@@ -0,0 +1,8 @@
+require "test_helper"
+
+class BooksControllerTest < ActionDispatch::IntegrationTest
+  test "should get index" do
```
Comment (8kristy) [inline test/controllers/books_controller_test.rb:4]: Might be worth playing around with mocking the DB and checking properties of the pages, e.g. does the response contain a particular book

Ideally we'll cover all methods and all edge cases within each one

File: test/fixtures/books.yml [hunk 20]
```diff
@@ -0,0 +1,11 @@
+# Read about fixtures at https://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
+
+# This model initially had no columns defined. If you add columns to the
+# model remove the "{}" from the fixture names and add the columns immediately
+# below each fixture, per the syntax in the comments below
+#
+one: {}
```
Comment (8kristy) [inline test/fixtures/books.yml:7]: I'm not too familiar with Ruby testing but I think this is used to create mock test data? https://thoughtbot.com/blog/we-need-to-talk-about-fixtures

File: test/models/book_test.rb [hunk 21]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
+class BookTest < ActiveSupport::TestCase
+  # test "the truth" do
```
Comment (8kristy) [inline test/models/book_test.rb:4]: Think it's worth testing the models too, this looks useful? https://semaphoreci.com/community/tutorials/how-to-test-rails-models-with-rspec

**Review by 8kristy — COMMENTED** [review]
Well done! There's a few nit picks about the spacing between methods, I left suggestions where I found them but would be good to run the formatter in your IDE to cover them all

Otherwise just a few general comments and I'd recommend to practice test writing :)

File: app/helpers/books_helper.rb (hunk 13 above)
Comment (liliannaszabo) [reply app/helpers/books_helper.rb:1]: Not really sure, these were automatically configured by Ruby when I created the models
